### PR TITLE
implement room service with error handling

### PR DESCRIPTION
**Pull request by mingxin-yu**
Implement room service

### CODE REVIEW

File: src/app/app.module.ts [hunk 1]
```diff
@@ -2,15 +2,25 @@ import { BrowserModule } from '@angular/platform-browser';
...
     BrowserModule,
-    AppRoutingModule
+    AppRoutingModule,
+    HttpClientModule,
+
+    // The HttpClientInMemoryWebApiModule module intercepts HTTP requests
+    // and returns simulated server responses.
+    // Remove it when a real server is ready to receive requests.
+    HttpClientInMemoryWebApiModule.forRoot(
+      InMemoryDataService, { dataEncapsulation: false }
```
Comment (ghost) [inline src/app/app.module.ts:22]: no space

Comment (ghost) [reply src/app/app.module.ts:22]: no space before and afer

File: src/app/room.service.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
+import { Injectable } from '@angular/core';
+import { HttpClient, HttpHeaders } from '@angular/common/http';
+
+import { Observable, of } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+
+import { Room } from './Room';
+
+
+@Injectable({ providedIn: 'root' })
```
Comment (ghost) [inline src/app/room.service.ts]: no space before and after

File: src/app/room.service.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+export class RoomService {
+
```
Comment (ghost) [inline src/app/room.service.ts]: remove blank line

File: src/app/room.service.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  private roomsUrl = 'api/rooms';  // URL to web api
+
+  httpOptions = {
+    headers: new HttpHeaders({ 'Content-Type': 'application/json' })
+  };
+
+  constructor(
+    private http: HttpClient) { }
```
Comment (ghost) [inline src/app/room.service.ts]: on same line with 19, remove space in {}

File: src/app/room.service.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import { Room } from './Room';
+
+
+@Injectable({ providedIn: 'root' })
+export class RoomService {
+
+  private roomsUrl = 'api/rooms';  // URL to web api
+
+  httpOptions = {
+    headers: new HttpHeaders({ 'Content-Type': 'application/json' })
```
Comment (ghost) [inline src/app/room.service.ts]: no space

File: src/app/room.service.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  /** GET room by id. Will 404 if id not found */
+  getRoom(id: number): Observable<Room> {
+    const url = `${this.roomsUrl}/${id}`;
+    return this.http.get<Room>(url).pipe(
+      catchError(this.handleError<Room>(`getRoom id=${id}`))
+    );
+  }
+
+  /* GET rooms whose name contains search term */
+  searchRooms(term: string): Observable<Room[]> {
```
Comment (ghost) [inline src/app/room.service.ts]: what is this method? Where is this used? Why did you write this?

File: src/app/room.service.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    if (!term.trim()) {
+      // if not search term, return empty room array.
+      return of([]);
+    }
+    return this.http.get<Room[]>(`${this.roomsUrl}/?name=${term}`).pipe(
+      catchError(this.handleError<Room[]>('searchRooms', []))
+    );
+  }
+
+  //////// Save methods //////////
```
Comment (ghost) [inline src/app/room.service.ts]: ?

File: src/app/room.service.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        catchError(this.handleError<Room[]>('getRooms', []))
+      );
+  }
+
+  /** GET room by id. Return `undefined` when id not found */
+  getRoomNo404<Data>(id: number): Observable<Room> {
+    const url = `${this.roomsUrl}/?id=${id}`;
+    return this.http.get<Room[]>(url)
+      .pipe(
+        map(rooms => rooms[0]), // returns a {0|1} element array
```
Comment (ghost) [inline src/app/room.service.ts]: put commente above the line

File: src/app/room.service.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+import { Observable, of } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+
+import { Room } from './Room';
+
+
+@Injectable({ providedIn: 'root' })
+export class RoomService {
+
+  private roomsUrl = 'api/rooms';  // URL to web api
```
Comment (ghost) [inline src/app/room.service.ts]: const private

Comment (ghost) [reply src/app/room.service.ts]: comment put above

File: src/app/room.service.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+  /** POST: add a new room to the server */
+  addRoom(room: Room): Observable<Room> {
+    return this.http.post<Room>(this.roomsUrl, room, this.httpOptions).pipe(
+      catchError(this.handleError<Room>('addRoom'))
+    );
+  }
+
+  /** DELETE: delete the room from the server */
+  deleteRoom(room: Room | number): Observable<Room> {
```
Comment (ghost) [inline src/app/room.service.ts]: This is not described in the design? why two arguments?

File: src/app/room.service.ts [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    const id = typeof room === 'number' ? room : room.id;
+    const url = `${this.roomsUrl}/${id}`;
+
+    return this.http.delete<Room>(url, this.httpOptions).pipe(
+      catchError(this.handleError<Room>('deleteRoom'))
+    );
+  }
+
+  /** PUT: update the room on the server */
```
Comment (ghost) [inline src/app/room.service.ts]: comments shouldn't describe exactly the method name.

File: src/app/app.module.ts (hunk 1 above)
Comment (mingxin-yu) [reply src/app/app.module.ts:22]: Ignore

File: src/app/room.service.ts (hunk 2 above)
Comment (mingxin-yu) [reply src/app/room.service.ts]: Ignore

File: src/app/room.service.ts (hunk 3 above)
Comment (mingxin-yu) [reply src/app/room.service.ts]: Done

File: src/app/room.service.ts (hunk 4 above)
Comment (mingxin-yu) [reply src/app/room.service.ts]: Done

File: src/app/room.service.ts (hunk 5 above)
Comment (mingxin-yu) [reply src/app/room.service.ts]: Ignore

File: src/app/room.service.ts (hunk 7 above)
Comment (mingxin-yu) [reply src/app/room.service.ts]: Deleted

File: src/app/room.service.ts (hunk 8 above)
Comment (mingxin-yu) [reply src/app/room.service.ts]: Done

File: src/app/room.service.ts (hunk 9 above)
Comment (mingxin-yu) [reply src/app/room.service.ts]: Done

File: src/app/room.service.ts (hunk 10 above)
Comment (mingxin-yu) [reply src/app/room.service.ts]: Changed

File: src/app/room.service.ts (hunk 11 above)
Comment (mingxin-yu) [reply src/app/room.service.ts]: Deleted

**Review by ghost — APPROVED** [review]

**Pull request merged**